src/drivers/driver_zephyr.c

```diff
@@ -0,0 +1,540 @@
+{
+	wpa_supplicant_event(if_ctx->supp_if_ctx, EVENT_SCAN_STARTED, NULL);
+}
+
+void wpa_drv_zep_event_proc_scan_done(struct zep_drv_if_ctx *if_ctx,
+				      union wpa_event_data *event)
+{
+	eloop_cancel_timeout(wpa_drv_zep_scan_timeout, if_ctx,
+			     if_ctx->supp_if_ctx);
+
+	wpa_supplicant_event(if_ctx->supp_if_ctx, EVENT_SCAN_RESULTS, event);
+}
+
+void wpa_drv_zep_event_proc_scan_res(struct zep_drv_if_ctx *if_ctx,
+				     struct wpa_scan_res *r, bool more_res)
+{
+	struct wpa_scan_res **tmp = NULL;
+
+	tmp = os_realloc_array(if_ctx->scan_res2->res,
+			       if_ctx->scan_res2->num + 1,
+			       sizeof(struct wpa_scan_res *));
+
+	if (!tmp) {
+		os_free(r);
+		return;
+	}
+
+	tmp[if_ctx->scan_res2->num++] = r;
+
+	if_ctx->scan_res2->res = tmp;
+
+	if_ctx->scan_res2_get_in_prog = more_res;
+}
+
+void wpa_drv_zep_event_proc_auth_resp(struct zep_drv_if_ctx *if_ctx,
+				      union wpa_event_data *event)
+{
+	wpa_supplicant_event(if_ctx->supp_if_ctx, EVENT_AUTH, event);
+}
+
+void wpa_drv_zep_event_proc_assoc_resp(struct zep_drv_if_ctx *if_ctx,
+				       union wpa_event_data *event,
+				       unsigned int status)
+{
+	if (status != WLAN_STATUS_SUCCESS) {
+		wpa_supplicant_event(if_ctx->supp_if_ctx, EVENT_ASSOC_REJECT,
+				     event);
+	} else {
+		if_ctx->associated = true;
+		os_memcpy(if_ctx->bssid, event->assoc_info.addr, ETH_ALEN);
+
+		wpa_supplicant_event(if_ctx->supp_if_ctx, EVENT_ASSOC, event);
+	}
+}
+
+void wpa_drv_zep_event_proc_deauth(struct zep_drv_if_ctx *if_ctx,
+				   union wpa_event_data *event)
+{
+	wpa_supplicant_event(if_ctx->supp_if_ctx, EVENT_DEAUTH, event);
+}
+
+void wpa_drv_zep_event_proc_disassoc(struct zep_drv_if_ctx *if_ctx,
+				     union wpa_event_data *event)
+{
+	wpa_supplicant_event(if_ctx->supp_if_ctx, EVENT_DISASSOC, event);
+}
+
+static void *wpa_drv_zep_global_init(void *ctx)
+{
+	struct zep_drv_ctx *drv_ctx = NULL;
+
+	drv_ctx = os_zalloc(sizeof(*drv_ctx));
+
+	if (!drv_ctx) {
+		return NULL;
+	}
+
+	drv_ctx->supp_ctx = ctx;
+
+	return drv_ctx;
+}
+
+static void wpa_drv_zep_global_deinit(void *priv)
+{
+	struct zep_drv_ctx *drv_ctx = priv;
+
+	if (!drv_ctx) {
+		return;
+	}
+
+	os_free(drv_ctx);
+}
+
+/**
+ * wpa_driver_zep_init - Initialize Zephyr driver interface
+ * @ctx: Context to be used when calling wpa_supplicant functions,
+ *       e.g., wpa_supplicant_event()
+ * @ifname: Interface name, e.g., wlan0
+ * @global_priv: private driver global data from global_init()
+ *
+ * Returns: Pointer to private data, %NULL on failure
+ */
+static void *wpa_drv_zep_init(void *ctx, const char *ifname, void *global_priv)
+{
+	struct zep_drv_if_ctx *if_ctx = NULL;
+	const struct zep_wpa_supp_dev_ops *dev_ops = NULL;
+	struct zep_wpa_supp_dev_callbk_fns callbk_fns;
+
+	if_ctx = os_zalloc(sizeof(*if_ctx));
+
+	if (if_ctx == NULL) {
+		return NULL;
+	}
+
+	if_ctx->supp_if_ctx = ctx;
+
+	if_ctx->dev_ctx = net_if_get_device(net_if_get_default());
```

Comment:
@rlubos Is this appropriate? What if there are multiple interfaces, e.g., Thread and WiFi? Will this still work?

Reply:
Good point, this won't be reliable in case of multiple network interfaces. We should likely use `device_get_binding()` here, in Zephyr interface names corresponds to the network device name.

Reply:
Fixed to use proper interface name